docs/Overview.md

```diff
@@ -1,96 +1,246 @@
+        case "String":
+            e = &StringExecutor{...}
+        case "AST":
+            e = &ASTExecutor{...}
+        case "SemanticGraph":
+            // the interpreter is an Executor!
+            e = interpreter.NewInterpreter()
+        default:
+            return nil, fmt.Errorf("unknown language: %v", l)
+    }
+    
+    q := ... // new query
+    ec := ... // create execution context
+    err := e.Execute(q, ec)
+    return q, err
 }
+```
 
-// Predicate filters data.
-type Predicate interface {}
+The `StringExecutor` does not actually know how to execute a script, but it uses a compiler to translate the script into its
+AST representation and delegate to an `ASTExecutor` in turn:
+
+```go
+func (e *StringExecutor) Execute(q Query, ec ExecutionContext) error {
+    if ln := q.Representation().Language().Name(); ln != "String" {
+        return fmt.Errorf("String executor cannot execute %s", ln)
+    }
+    
+    astc := &ASTCompiler{...}
+    newR, err := astc.Compile(q.Representation())
+    if err != nil {
+        return fmt.Errorf("cannot compile")
+    }
+    q.SetRepresentation(newR) // substitute Representation
+    aste, err:= &ASTExecutor{...}
+    return se.Execute(q, ec) // forward the execution context!
+}
+```
+
+The `ASTExecutor` does not actually know how to execute an AST, but it uses a compiler to translate the AST into a
+semantic graph and delegate to a `SemanticExecutor` in turn:
+
+```go
+func (e *ASTExecutor) Execute(q Query, ec ExecutionContext) error {
+    if ln := q.Representation().Language().Name(); ln != "AST" {
+        return fmt.Errorf("AST executor cannot execute %s", ln)
+    }
+    
+    sgc := &SemanticGraphCompiler{...}
+    newR, err := sgc.Compile(q.Representation())
+    if err != nil {
+        return fmt.Errorf("cannot compile")
+    }
+    q.SetRepresentation(newR) // substitute new Representation
+    // the interpreter is an Executor!
+    itrp := interpreter.NewInterpreter()
+    return itrp.Execute(q, ec) // forward the execution context!
+}
+```
+
+The `Interpreter` is responsible for executing semantic graphs:
+
+```go
+func (i *Interpreter) Execute(q Query, ec ExecutionContext) error {
+    if ln := q.Representation().Language().Name(); ln != "SemanticGraph" {
+        return fmt.Errorf("Interpreter cannot execute %s", ln)
+    }
+    
+    itrp := interpreter.NewInterpreter()
+    universe := ec.Bindings()
+    semPkg := q.Representation().(*semantic.Package)
+    // pass the execution context!
+    results, err := itrp.Eval(semPkg, universe, ec)
+    if err != nil {
+        return err
+    }
+    // ... add results the query (see below).
+    return nil
+}
+```
+
+While evaluating, the interpreter updates the scope of execution by visiting the semantic graph. It performs
+additions, subtractions, comparisons, etc., variable assignments, imports, and data manipulation functions. 
+The interpreter can further compile the semantic graph to an IR and to an
```

Comment:
Just to drive the point home. The interpreter is an executor. And not just any executor, it is the Flux Virtual Machine. It executes a typed semantic graph and produces a query result. Because it executes queries it must embed a query engine. And because the query engine executes query plans, the interpreter must also embed a compiler that targets the query engine. That's how I see it at least.